libs/wingsdk/src/shared/sandbox.ts

```diff
@@ -109,55 +45,121 @@ export class Sandbox {
     }
   }
 
-  private loadBundleOnce() {
-    if (this.loaded) {
-      return;
-    }
-
-    if (!this.code) {
-      throw new Error("Bundle not created yet - please report this as a bug");
-    }
-
-    // this will add stuff to the "exports" object within our context
-    vm.runInContext(this.code!, this.context, {
-      filename: this.entrypoint,
-    });
-
-    this.loaded = true;
-  }
-
   public async call(fn: string, ...args: any[]): Promise<any> {
     // wait for the bundle to finish creation
     await this.createBundlePromise;
 
-    // load the bundle into context on the first run
-    // we don't do this earlier because bundled code may have side effects
-    // and we want to simulate that a function is "cold" on the first run
-    this.loadBundleOnce();
-
-    return new Promise(($resolve, $reject) => {
-      const cleanup = () => {
-        delete this.context.$resolve;
-        delete this.context.$reject;
-      };
-
-      this.context.$resolve = (value: any) => {
-        cleanup();
-        $resolve(value);
-      };
+    // emit an explicit error when trying to access __dirname and __filename because we cannot
+    // resolve these when bundling (this is true both for simulator and the cloud since we are
+    // bundling there as well).
+    const shim = `
+const { parentPort: $parent } = require('worker_threads');
+const $resolve = (value) => $parent.postMessage({ type: 'resolve', value });
+const $reject = (reason) => $parent.postMessage({ type: 'reject', reason });
+console.log = (...args) => $parent.postMessage({ type: 'log', args });
+console.debug = (...args) => $parent.postMessage({ type: 'debug', args });
+console.info = (...args) => $parent.postMessage({ type: 'info', args });
+console.warn = (...args) => $parent.postMessage({ type: 'warn', args });
+console.error = (...args) => $parent.postMessage({ type: 'error', args });
+process.exit = (exitCode) => $parent.postMessage({ type: 'exit', exitCode });
+Object.defineProperty(globalThis, "__dirname", {
+  get: () => { throw new Error("__dirname cannot be used within bundled cloud functions"); },
+});
+Object.defineProperty(globalThis, "__filename", {
+  get: () => { throw new Error("__filename cannot be used within bundled cloud functions"); },
+});
+${this.code}
+$parent.on('message', async (message) => {
+  try {
+    const result = await exports[message.fn](...message.payload);
+    $resolve(result);
+  } catch (error) {
+    $reject(error);
+  }
+});
+      `;
+    console.error(shim);
+
+    // currently, a fresh worker is used for every invocation
+    // it could be better to keep the worker alive and reuse it
+    // but this requires additional work to make sure logs between invocations
+    // are not mixed up, and timeouts are handled correctly
+    const worker = new Worker(shim, {
+      env: this.options.env,
+      eval: true,
+    });
```

Comment:
To fix sourcemaps, a more typical way to do this is to use a preload script instead of an eval'd shim:

```ts
const worker = new Worker(this.entrypoint, {
  env: this.options.env,
  // create a local file called sandbox-shim.ts with all the extra stuff needed
  // This file will be executed first, then the entrypoint
  execArgv: ["-r", require.resolve("./sandbox-shim")],
});
```